daemon/src/common/Profiler.cpp

```diff
@@ -0,0 +1,184 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#include "Profiler.h"
+#include "../engine/framework/LogSystem.h"
+
+#include "../engine/framework/CommandSystem.h"
+#include "../engine/framework/BaseCommands.h"
+#include "FileSystem.h"
+#include <fstream>
+
+
+namespace Profiler{
+
+    /*
+         * Function execution times are saved in the samples vector, every element
+         * contains the name of the function, the type of sample (function start,
+         * function end, and if there's a new frame), and a timestamp.
+         */
+    std::vector <Profiler::Point>                   samples;
+    Sys::SteadyClock::time_point                    startTime;
+    bool                                            enabled    = false; ///TODO those bools are silly
+    bool                                            nextFrame  = false;
+    bool                                            shouldStop = false;
+
+
+    /// Called at the end of every frame
+    void Update(){
+
+        if(nextFrame){
+            enabled=true;
+            shouldStop=true;
+            nextFrame=false;
+            startTime   = Sys::SteadyClock::now();
+            return;
+        }
+
+        if(shouldStop){
+            enabled=false;
+            shouldStop=false;
+            Stop();
+            return;
+        }
+
+        if(!enabled)
+            return;
+        samples.push_back(Point(FRAME,""));
+    }
+
+
+    /// Point saves the time it's constructed
+    Point::Point(Type type, const char *label=""): type(type), label(label){
+        auto timeElapsed = Sys::SteadyClock::now() - startTime;
+        time = std::chrono::duration_cast < std::chrono::microseconds > (timeElapsed).count();
+    }
+
+
+    /// At the Beginning and End of every function
+
+    Profile::Profile( const char * label ): label( label ){
+        if(!enabled)
+            return;
+        samples.push_back( Point( START, label ) );
+    }
+
+    Profile::~Profile(){
+        if(!enabled)
+            return;
+        samples.push_back( Point( END, label ) );
+    }
+
+
+    void Stop(){
+        enabled=false;
+
+        auto file = FS::HomePath::OpenWrite("profile.log"); ///TODO add timestamp to file
```

Comment:
I was concerned about speed too, writing to the file at every function call is out of discussion, it may be feasible at every frame but it would still have a performance impact...